random: generate random seed at compile-time

nmeum opened the pull request:
### Contribution description

If the device does not support run-time random number generation RIOT currently uses a hardcoded seed. I discovered this while working on #13253. This PR uses the technique from #13119 to randomize the `RANDOM_DEFAULT_SEED` on each build. Currently, this technique is used unconditionally. Alternatively, it would be possible to only use it if run-time random number generation is not supported. However, the current code has some other issues, for instance `puf_sram` may not necessarily return a fresh value. For this reason, I decided to combine the compile-time generated seed with a run-time generated one (if available) using a multiply operation. **I am not a cryptographic expert but this shouldn't have a negative impact on the entropy**.

### Testing procedure

I tested this with `tests/rng`. I also disabled the `periph_cpuid` and `periph_hwrng` feature for `BOARD=native` to make sure the fallback code works.

### Issues/PRs references

* #13253 
* #13119

nmeum commented on sys/random/init.c:
```diff
@@ -0,0 +1,48 @@
+#else
+    LOG_WARNING("random: USING COMPILE-TIME GENERATED SEED!\n");
+#endif
+    seed *= RANDOM_SEED_DEFAULT;
```
Not sure if multiplication is really the best mathematical operation to use here. Maybe it makes sense to use XOR instead? Any preferences on this?

miri64 replied:
Not a crypto crack either, but intuitively, XOR (which "shakes up the bits") seems more suitable than multiplication (which if it is "accidentally" of two can lead to a lot of LSB zeros).

PeterKietzmann replied:
This is only a compile-static fallback, right? I would rather leave it in the `#else` case and with that I don't see the need for combing the static seed with `seed`  anyway.  In the long announced and yet not seen entropy module (shame on me) that I plan to propose, a static input will be enabled as entropy source which will then (ideally) be hashed together with other entropy sources.

nmeum replied:
Is not strictly necessary to combine the seed with `RANDOM_SEED_DEFAULT` unconditionally. Preferably I would not want to recompile the file using `RANDOM_SEED_DEFAULT` on each make invocation if some other method of generating random numbers is available though I didn't find any easy way to detect whether or not this is the case from the `Makefile`. If there is such a method let me know, in that case it would be possible to only mark the file as `.PHONY` if no run-time random number generation method is available.

nmeum commented:
> I still think that this doesn't solve the problem for devices without proper RNG. It'll feel like, for us developers, but real deployments™ are stuck with the same random seed, giving a false sense of security.

I absolutely agree that it doesn't solve the problem, I never intended to claim that it did. However, I think that this is an improvement over the current fallback code which uses a simple hardcoded value. I would also be fine with removing the `RANDOM_SEED_DEFAULT` fallback in favor of requiring `puf_sram` or some other method. Also keep in mind that this method improves run-time methods such as `puf_sram` by combining them with a compile-time value.

PeterKietzmann commented on sys/random/init.c:
```diff
@@ -0,0 +1,48 @@
+#include "luid.h"
+#endif
+
+#define ENABLE_DEBUG (1)
```
We disable this by default.

nmeum replied:
Fixed, thanks.

PeterKietzmann commented on tests/gnrc_ipv6_nib/Makefile.ci:
```diff
@@ -11,6 +11,7 @@ BOARD_INSUFFICIENT_MEMORY := \
     nucleo-l031k6 \
     stm32f030f4-demo \
     telosb \
+    msb-430 \
```
Doesn't relate to this PR?

nmeum replied:
Yes, seems the PR increased the binary size of the `random` library a bit which caused it to exceed the available memory on this board.

nmeum commented:
> @nmeum thanks for the contribution! I'd like to discuss a little bit more but in general I'm not against a randomized compile-static input. However, I don't see why simply changing `RANDOM_SEED_DEFAULT` during compilation isn't enough.

Changing `RANDOM_SEED_DEFAULT` is enough. I just moved the code using it to a separate file to make sure recompiling this file on each `make` invocation is fast. Probably, doesn't make a big difference so if desired I can just revert that change.

PeterKietzmann commented:
Again to this PR:
- I'm fine introducing *randhex.py* for compile static generation of a `RANDOM_SEED_DEFAULT`.
- Moving the initialization function to a separate file is fine with me too.
- I still don't like the unconditional `seed *= RANDOM_SEED_DEFAULT` in presence of a "proper" seed. If I interpret @kaspar030's comment and its reactions properly, others see it in the same way. In my opinion, to add such a "personalization component", we should combine it with a decent derivation function which is out of scope of this PR.

> Preferably I would not want to recompile the file using RANDOM_SEED_DEFAULT on each make invocation if some other method of generating random numbers is available

If an other source would be available on the pĺatfrom that you build for, why do you "need" `RANDOM_SEED_DEFAULT`? 

> For instance, on BOARD=bluepill I currently still end-up with the same ephemeral port

Unfortunately, I don't have that board at hand. Are you saying that the `puf_sram` module doesn't work on this platform?

nmeum commented:
I agree which everything you are saying, it just didn't get around to making the necessary adjustments yet.

nmeum commented:
Added some changes which only use a compile-time generated seed as a fallback and don't instrument other seeds with an additional compile-time value. I also added some magic to the Makefile to ensure that `init.c` is only marked as `.PHONY` if the fallback code is actually in use. Let me know what you think.

PeterKietzmann commented on sys/include/random.h:
```diff
@@ -34,14 +34,6 @@
- *          failed
- */
-#define RANDOM_SEED_DEFAULT (1)
-#endif
```
I'm not sure if we should simply remove this block.
(i) It prevents users to set their own fallback seed. It is a relevant use case? Yes, for debugging purposes. However, in that case, we would want a static value over all other seed sources anyway, which was never the case. This requires (minimal) additional configuration effort **but*  it could justify to always randomize the static fallback... 
(ii) It removes documentation of that parameter.

What do you think @nmeum?

nmeum replied:
Don't have any strong opinions on this since it's easy to implement and doesn't have any obvious disadvantages. I am fine with either option.

nmeum replied:
Though one obvious disadvantage would be that `RANDOM_SEED_DEFAULT` would be set to a constant value without anyone noticing.

PeterKietzmann replied:
I'm wondering if anyone has ever even used this parameter? The two options that I see here are as follows:
- Simply remove the parameter because it is "nonsense". It is not "nonsense" for debugging purposes, but a dedicated debug mode would be implemented differently anyway.
- Add a switch so that either `RANDOM_SEED_DEFAULT` or *randhex* can be used. 

Opinions @nmeum, @leandrolanzieri?

nmeum replied:
> I'm wondering if anyone has ever even used this parameter?

I think its unlikely, especially since the parameter is only used if no other seed generation facility is available. I think the interesting question is: In which case does it make sense to supply a hardcoded seed for random number generation. Debugging is certainly a scenario where this might be useful but in that case `RANDOM_SEED_DEFAULT` must take precedence over the other facilities. For debugging purposes it might even be sufficient to simply modify  the `init.c` source. 

IMHO we could remove `RANDOM_SEED_DEFAULT` entirely for now, and maybe add a "dedicated debug mode" if it turns out to be needed or is explicitly requested by developers?

PeterKietzmann replied:
Yes, that is what I meant with "dedicated debug mode would be implemented differently anyway" so I agree with you. Let's wait 1-2 days with merging so others still have time to state their opinions.

In the meantime, how about we add a compile time warning for the case that a proper entropy source is missing (like [here](https://github.com/RIOT-OS/RIOT/blob/master/sys/puf_sram/Makefile#L1))?

nmeum replied:
> In the meantime, how about we add a compile time warning for the case that a proper entropy source is missing (like here)?

Sounds good, will do that tomorrow! :blush:

nmeum replied:
Added a warning to the makefile.

PeterKietzmann commented on sys/random/Makefile:
```diff
@@ -11,4 +14,28 @@ ifneq (,$(filter prng_tinymt32,$(USEMODULE)))
+endif
+ifneq (,$(filter periph_cpuid,$(FEATURES_PROVIDED)))
+  HAS_RNG = 1
+endif
```
That looks bad although I believe there is currently no nicer way. @leandrolanzieri can you have a look?

nmeum replied:
Also fails the `buildsystem_sanity_check`:

	Invalid build system patterns found by ./dist/tools/buildsystem_sanity_check/check.sh:
	Modules should not check the content of FEATURES_PROVIDED/_REQUIRED/OPTIONAL
		sys/random/Makefile:ifneq (,$(filter periph_hwrng,$(FEATURES_PROVIDED)))
		sys/random/Makefile:ifneq (,$(filter periph_cpuid,$(FEATURES_PROVIDED)))

leandrolanzieri replied:
One should check for `FEATURES_USED`

nmeum commented:
> Unfortunately, I don't have that board at hand. Are you saying that the puf_sram module doesn't work on this platform?

I didn't test the `puf_sram` module explicitly. All I did was use an unmodified version of `examples/asymcute_mqttsn` with #13253 and `BOARD=bluepill`. While doing so, I noticed that it used the same ephemeral port on each boot. Revisiting this now it seems that this was due to the fact that is uses `luid_get` for the initial random seed and it seems to me that `luid_get` just returns a derivation of the serial which is obviously the same on each boot. If so, it could IMHO be useful to combine a compile-time generated value with the value returned by `luid_get`.

Regarding `puf_sram`, it seems to be me that it has to be enabled explicitly by adding `USEMODULE += puf_sram` to the application Makefile, right? If so: It might make sense to enable it by default (if available) to improve the default seed generation of the `random` module?

> @kaspar030 raised concerns about giving a false sense of security.

I think the entire random module currently gives a false sense of security. The problems regarding random number generations, the methods available, and their deficiencies (see for instance my observations regarding `luid_get` above) should IMHO be documented somewhere. If such a document is already available please let me know.

Also: `luid_get` seems to use `0x23` as a fallback value if `CPUID_LEN` is not defined. Obviously, this has the same disadvantages as a constant default seed (which this PR attempts to remove):

https://github.com/RIOT-OS/RIOT/blob/38e94d09dbc0d299e555242e44cc7069d089d9b0/sys/luid/luid.c#L35-L45

miri64 commented:
> Also: `luid_get` seems to use `0x23` (`LUID_BACKUP_SEED`) as a fallback value if `CPUID_LEN` is zero? Obviously, this has the same disadvantages as a constant default seed (which this PR attempts to remove):

Regarding address generation this is a two cut-sword. With multiple nodes you want to of course have different addresses, however, the same address at reflash/boot-up/... is also desirable.

PeterKietzmann commented:
> While doing so, I noticed that it used the same ephemeral port on each boot. Revisiting this now it seems that this was due to the fact that is uses luid_get for the initial random seed

Seeding with a static CPUID is bad, that's for sure and thus, a randomized compile static input is the last hope to create some derivation in the absence of other entropy sources. That's why I'm not against this PR. However, the compile static input does not help you on a reboot.

> Regarding puf_sram, it seems to be me that it has to be enabled explicitly by adding USEMODULE += puf_sram to the application Makefile, right? If so: It might make sense to enable it by default

We considered it once, but IMO this is too dangerous. (i) Its properties should be [evaluated beforehand](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/puf_sram.h#L32) and (ii) it requires a power-off cycle. I got myself a bluepill and may give it a try soon. However, on a nucleo-f103 the mechanism works well.

> Regarding address generation this is a two cut-sword. With multiple nodes you want to of course have different addresses, however, the same address at reflash/boot-up/... is also desirable.

I basically agree with this statement but I fail to see how it relates to seed generation.

nmeum commented:
> However, the compile static input does not help you on a reboot.

I absolutely agree, the deficiencies of this mechanism should be clearly communicated (as should the deficiencies of the other mechanisms implemented so far) . This PR does not solve the "random number generation"-issue, I just intend it to be an additional source of entropy.

> I got myself a bluepill and may give it a try soon. However, on a nucleo-f103 the mechanism works well.

I don't doubt that it works on the bluepill, I just didn't know about it beforehand and therefore did not use it for my ephemeral port tests.

miri64 commented:
> > Regarding address generation this is a two cut-sword. With multiple nodes you want to of course have different addresses, however, the same address at reflash/boot-up/... is also desirable.
> 
> I basically agree with this statement but I fail to see how it relates to seed generation.

`luid_get()` (which is used for address generation) was brought up. That's why I wanted to clarify the static nature of the seed for that function.

leandrolanzieri commented on sys/random/Makefile:
```diff
@@ -11,4 +14,28 @@ ifneq (,$(filter prng_tinymt32,$(USEMODULE)))
+  RANDOM_SEED := 0x1
+endif
+
+HAS_RNG := 0
+ifneq (,$(filter puf_sram,$(USEMODULE)))
+  HAS_RNG = 1
+endif
+ifneq (,$(filter periph_hwrng,$(FEATURES_PROVIDED)))
+  HAS_RNG = 1
+endif
+ifneq (,$(filter periph_cpuid,$(FEATURES_PROVIDED)))
+  HAS_RNG = 1
+endif
+
+ifeq (0,$(HAS_RNG))
+  CFLAGS += -DRANDOM_SEED_DEFAULT=$(RANDOM_SEED)U
+  .PHONY: $(BINDIR)/$(MODULE)/init.o
+endif
```
```suggestion
ifneq (,$(filter puf_sram,$(USEMODULE)))
  HAS_RNG = 1
else ifneq (,$(filter periph_hwrng periph_cpuid,$(FEATURES_PROVIDED)))
  HAS_RNG = 1
endif

ifndef HAS_RNG
  CFLAGS += -DRANDOM_SEED_DEFAULT=$(RANDOM_SEED)U
  .PHONY: $(BINDIR)/$(MODULE)/init.o
endif
```

PeterKietzmann commented:
Regarding SRAM seeding on the bluepill, it works well. I ran the [test application](https://github.com/RIOT-OS/RIOT/tree/master/tests/puf_sram) which led to following results:

```
Number of seeds: 500       
Seed length    : 32 Bit   
Abs. Entropy   : 30.13 Bit   
Rel. Entropy   : 94.17 perc. 
```

1. These results are obtained while re-powering  the board. You can achieve the same effect with a low-power cycle without RAM retention. However, you need at least one power cycle after flashing the device.
2. Soft resets will change the seed, but don't expect fresh entropy then (see #13137 for reference).

nmeum commented:
> Regarding SRAM seeding on the bluepill, it works well.

Good to know, thanks for trying! Unless we decide to enable `puf_sram` in favor of the other methods currently used for seeding (which [does not seem to be an option](https://github.com/RIOT-OS/RIOT/pull/13255#issuecomment-610364532)) this does not really change the nature of this PR.

nmeum commented on sys/random/Makefile:
```diff
@@ -11,4 +14,28 @@ ifneq (,$(filter prng_tinymt32,$(USEMODULE)))
+endif
+
+# This has to be kept in sync with sys/random/init.c
+ifneq (,$(filter puf_sram periph_hwrng periph_cpuid,$(USEMODULE)))
```
Just as a heads up: Changed this check, as it seems checking `FEATURES_USED` is not sufficient, instead `USEMODULE` needs to be used for this check.

nmeum commented on sys/random/Makefile:
```diff
@@ -11,4 +14,28 @@ ifneq (,$(filter prng_tinymt32,$(USEMODULE)))
+# generating an initial seed, generate a random value at compile-time.
+# Also mark the file using it as PHONY to use a new seed on each build.
+ifndef HAS_RNG
+  $(warning Attention! The random module is using a compile-time seed. \
+    The same seed will be used for the same firmware image.)
```
If someone has an idea how to word this better, i.e. explain in the shortest way possible what's going on, let me know.

PeterKietzmann replied:
Actually, seeing this, I'm wondering if we should add this kind of message for the CPUID seeder too... Maybe a shared message if static seeds are involved? On the other had, we cannot add compile warnings everywhere :). What do you think?

nmeum replied:
I think the CPUID seeder needs to be improved in general. I already laid out [my issues with the CPUID seeder and random ephemeral ports](https://github.com/RIOT-OS/RIOT/pull/13255#issuecomment-610332381). Maybe combine the CPUID seeder with the compile-time value or stop using it all together? I don't know. In any case, the CPUID seeder has the deficiency that one always ends up with the same seed on the same board. That should definitely be documented somewhere but maybe changes to the CPUID seeder should be discussed in a separate issue / PR?

PeterKietzmann commented:
I've tested this this PR as follows:
- Compile for a z1 board: prints the new warning
- Compile & execute on native: nothing changes. Defaults to HWRNG.
- Compile & execute on nucleo-f410rb:  nothing changes. Defaults to HWRNG.
- Remove HWRNG from `FEATURES FPROVIDED`, compile & execute on nucleo-f410rb: selects CPUID
- Removed HWRNG & CPUID from `FEATURES FPROVIDED`, compile & execute on nucleo-f410rb: prints the new warning. The seed value does not change on reset but on re-compilation.
- Added `USEMODULE+=puf_sram`, removed HWRNG & CPUID from `FEATURES FPROVIDED`, compile & execute on nucleo-f410rb: uses `puf_sram` witout printing the warning

-> everything works as expected

PeterKietzmann commented:
Ping @kaspar030

benpicco commented on sys/random/Makefile:
```diff
@@ -11,4 +14,28 @@ ifneq (,$(filter prng_tinymt32,$(USEMODULE)))
   DIRS += tinymt32
 endif
 
+ifeq (,$(RIOT_CI_BUILD))
+  # random seed value newly generated on each build
+  RANDOM_SEED := $(shell $(RIOTTOOLS)/randhex/randhex.py 32)
```
Tbh this feature looks like a huge foot gun.
It will provide randomness during development, but once deployed all boards will behave the same.

If this feature is needed for a manual deployment where each node is flashed with a new build, it could be enabled through a pseudo-module. (But even then it would be much better to store the seed in a separate config region on the flash or EEPROM).

But making it default gives a false sense of randomness.

e.g. a user uses random() to decide when nodes should send something, during development with heterogeneous firmware versions this might work as expected.

When deployed all boards will send at the same time.

Can we instead work on improving the random API?

 - one 'true' random function (can be seeded by `puf_sram`, hwrng, …) that is not stable across reboots
 - instantiate randomness (seeded by CPU_ID or application) that can be used to generate 'stable' (across reboots) IDs or application behavior.

kaspar030 replied:
100% my thoughts.

Could we compromise on making this non-default?

miri64 replied:
Ping @nmeum, please comment.

nmeum replied:
I think I commented on this deficiency in this PR already. I don't really have the time to work on this anyhow anymore. I wouldn't mind if this just gets closed, the issues with the random module laid out in this PR still stand though.

PeterKietzmann commented:
> Tbh this feature looks like a huge foot gun.
It will provide randomness during development, but once deployed all boards will behave the same.

> But making it default gives a false sense of randomness.

So does [setting it to `0x1`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/random.h#L42). Additionally, there is a  [compile time warning](https://github.com/RIOT-OS/RIOT/pull/13255/files#diff-437ac60cf858166a11a50640fe808948R34) now. Varying seeds between builds and the warning improve the current situation. It is obvious that this mode should not be used for a real deployment.

> instantiable randomness (seeded by CPU_ID or application) that can be used to generate 'stable' (across reboots) IDs or application behavior.

CPUIDs are not always in place (very bad seeds anyway). Reseeding can be done by the application already, however, PRNGs need to run before your application code begins.

> Could we compromise on making this non-default?

Like, introducing a pseudomodule as suggested by @benpicco?  I wouldn't mind. As indicated before, this mode should only be used for debugging and development anyway...

miri64 commented:
Ping @nmeum
